docs/doc/60-contributing/03-rfcs/20220512-versioned-config.md

```diff
@@ -0,0 +1,141 @@
+---
+title: Versioned Config
+description: RFC for versioned config
+---
+
+- RFC PR: [datafuselabs/databend#5324](https://github.com/datafuselabs/databend/pull/5324)
+- Tracking Issue: [datafuselabs/databend#0000](https://github.com/datafuselabs/databend/issues/0000)
+
+# Summary
+
+The versioned configuration will allow us to iterate quickly while avoiding breaking the environment.
+
+# Motivation
+
+While early birds are starting to deploy databend by themselves, it's time for us to establish some contracts between users. We should allow users to upgrade their deployments without breaking backward compatibility. In this RFC, we will focus on `config`.
+
+`config` I mentioned here including:
+
+- config file that is read by `databend-query` and `databend-meta`.
+- config env that read by `databend-query` and `databend-meta`
+- application args that are accepted by `databend-query` and `databend-meta`
+- protobuf messaged that generated by `databend-query` (stored inside `databend-meta`)
+
+Out of scope:
+
+- Tools like `fuzz` and `metactl` are not covered by this RFC.
+- Command-line UX of `databend-query` and `databend-meta` is another topic. We will not cover it in this RFC.
+- Config input/output by SQL/HTTP Rest API are not covered (for example, the output of table `system.config`)
+
+> For convenience, I will use `databend` to refer to `databend-query` and `databend-meta`.
+
+With this RFC, our users will upgrade their deployments without breaking. Old configs should always work along with the new implementations.
+
+# Guide-level explanation
+
+With this RFC, users will have a new field called `version` at the top of the config file. The value of the version is an integer and defaults to `0`. If the version is not specified, we will load it as` 0`.
+
+All config files, env, and args will be supported:
+
+- config file: `version=42`
+- config env: `export CONFIG_VERSION=42`
+- args: `--config-version=42`
```

Comment:
the versioned config approach is commonly seen in many projects, but `--version` args approach is very quite familiar, can we have some similar examples in the other projects?

Reply:
To be honest, I didn't find a project accepting args like `config-version` :smile_cat: 

`--version` is different from `--config-version` we used here. `--version` has often been used as a flag which prints app versions and `--config-version` intends to specify the config version we are trying to load.

Although we support the usage like:

```shell
databend-query -c ./old-version.toml --config-version=42
```

But I think we don't need to mention the behavior inside docs. Because the args parser always uses the latest version (

Reply:
> Because the args parser always uses the latest version (

Well, it's also possible to support different versions of args...

Reply:
how about taking the args in a different manner? IMO args is harder to envolve, and less chance to change. 

etcd marks the obsoleted args with a DEPRECATED flag, so it give users some freedom to transist his/her args in production: https://etcd.io/docs/v3.2/op-guide/configuration/

Reply:
oh, also in etcd, it has an env var called ETCDCTL_API=3 to envolve the args, FYI

```
export ETCDCTL_API=3
```

Reply:
As described in RFC, command UX is another topic, and I think we should discuss it in a new RFC. In this RFC, we will keep the current behavior that all config fields map to a unique arg.